src/utils/exec.c

```diff
@@ -733,13 +734,54 @@ gboolean bd_utils_init_prog_reporting (BDUtilsProgFunc new_prog_func, GError **e
     return TRUE;
 }
 
+/**
+ * bd_utils_init_prog_reporting_thread:
+ * @new_prog_func: (allow-none) (scope notified): progress reporting function to
+ *                                                use on current thread or %NULL
+ *                                                to reset to default or global
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether progress reporting was successfully initialized or not
+ */
+gboolean bd_utils_init_prog_reporting_thread (BDUtilsProgFunc new_prog_func, GError **error __attribute__((unused))) {
+    /* XXX: the error attribute will likely be used in the future when this
+       function gets more complicated */
+
+    thread_prog_func = new_prog_func;
+
+    return TRUE;
+}
+
+static void thread_progress_muted (guint64 task_id __attribute__((unused)), BDUtilsProgStatus status __attribute__((unused)), guint8 completion __attribute__((unused)), gchar *msg __attribute__((unused))) {
+    /* This function serves as a special value for the progress reporting
+     * function to detect that nothing is done here. If the client uses an own
+     * empty function then bd_utils_prog_reporting_initialized will return TRUE.
+     */
+}
+
+/**
+ * bd_utils_mute_prog_reporting_thread:
+ * @error: (out): place to store error (if any)
+ *
+ * Returns: whether progress reporting for the current thread was successfully
+ * deinitialized or not
+ */
+gboolean bd_utils_mute_prog_reporting_thread (GError **error __attribute__((unused))) {
+    /* XXX: the error attribute will likely be used in the future when this
+       function gets more complicated */
+
+    thread_prog_func = thread_progress_muted;
+
+    return TRUE;
+}
+
 /**
  * bd_utils_prog_reporting_initialized:
  *
  * Returns: TRUE if progress reporting has been initialized.
  */
 gboolean bd_utils_prog_reporting_initialized () {
-    return prog_func != NULL;
+    return (thread_prog_func != NULL || prog_func != NULL) && thread_prog_func != thread_progress_muted;
```

Comment:
I think explicitly muted thread progress reporting should be considered initialized.

Reply:
> So this means that e.g. fsck would try to use the reporting function and parse the output.

Oh that's a good point, we should avoid the parsing overhead.

Reply:
> Should muting and disabling be considered differently? Then I could implement both, once for local muting (still initialized) and once for disabling (not initialized anymore).

Yes, that's what I head in mind. I think disabling should be possible like before. However, I see a benefit in being able to mute just specific thread(s).